packages/react-core/src/components/DualListSelector/examples/ComposableTree.tsx

```diff
@@ -0,0 +1,344 @@
+import AngleDoubleLeftIcon from '@patternfly/react-icons/dist/esm/icons/angle-double-left-icon';
+import AngleLeftIcon from '@patternfly/react-icons/dist/esm/icons/angle-left-icon';
+import AngleDoubleRightIcon from '@patternfly/react-icons/dist/esm/icons/angle-double-right-icon';
+import AngleRightIcon from '@patternfly/react-icons/dist/esm/icons/angle-right-icon';
+
+interface FoodNode {
+  id: string;
+  text: string;
+  children?: FoodNode[];
+}
+
+interface ExampleProps {
+  data: FoodNode[];
+}
+
+export const ComposableDualListSelectorTree: React.FunctionComponent<ExampleProps> = ({ data }: ExampleProps) => {
+  const [checkedLeafIds, setCheckedLeafIds] = React.useState<string[]>([]);
+  const [chosenLeafIds, setChosenLeafIds] = React.useState<string[]>(['beans', 'beef', 'chicken', 'tofu']);
+  const [chosenFilter, setChosenFilter] = React.useState<string>('');
+  const [availableFilter, setAvailableFilter] = React.useState<string>('');
+  let hiddenChosen = [] as string[];
+  let hiddenAvailable = [] as string[];
+
+  // helper function to build memoized lists
+  const buildTextById = (node: FoodNode): { [key: string]: string } => {
+    let textById = {};
+    if (!node) {
+      return textById;
+    }
+    textById[node.id] = node.text;
+    if (node.children) {
+      node.children.forEach(child => {
+        textById = { ...textById, ...buildTextById(child) };
+      });
+    }
+    return textById;
+  };
+
+  // helper function to build memoized lists
+  const getDescendantLeafIds = (node: FoodNode): string[] => {
+    if (!node.children || !node.children.length) {
+      return [node.id];
+    } else {
+      let childrenIds = [];
+      node.children.forEach(child => {
+        childrenIds = [...childrenIds, ...getDescendantLeafIds(child)];
+      });
+      return childrenIds;
+    }
+  };
+
+  // helper function to build memoized lists
+  const getLeavesById = (node: FoodNode): { [key: string]: string[] } => {
+    let leavesById = {};
+    if (!node.children || !node.children.length) {
+      leavesById[node.id] = [node.id];
+    } else {
+      node.children.forEach(child => {
+        leavesById[node.id] = getDescendantLeafIds(node);
+        leavesById = { ...leavesById, ...getLeavesById(child) };
+      });
+    }
+    return leavesById;
+  };
+
+  // Builds a map of child leaf nodes by node id - memoized so that it only rebuilds the list if the data changes.
+  const { memoizedLeavesById, memoizedAllLeaves, memoizedNodeText } = React.useMemo(() => {
+    let leavesById = {};
+    let allLeaves = [];
+    let nodeTexts = {};
+    data.forEach(foodNode => {
+      nodeTexts = { ...nodeTexts, ...buildTextById(foodNode) };
+      leavesById = { ...leavesById, ...getLeavesById(foodNode) };
+      allLeaves = [...allLeaves, ...getDescendantLeafIds(foodNode)];
+    });
+    return {
+      memoizedLeavesById: leavesById,
+      memoizedAllLeaves: allLeaves,
+      memoizedNodeText: nodeTexts
+    };
+  }, [data]);
```

Comment:
Shouldn't this be in the form of:

```js
  const memoizeNodes = (data: FoodNode[]) => {
    let leavesById = {};
    let allLeaves = [];
    let nodeTexts = {};
    data.forEach(foodNode => {
      nodeTexts = { ...nodeTexts, ...buildTextById(foodNode) };
      leavesById = { ...leavesById, ...getLeavesById(foodNode) };
      allLeaves = [...allLeaves, ...getDescendantLeafIds(foodNode)];
    });
    return {
      memoizedLeavesById: leavesById,
      memoizedAllLeaves: allLeaves,
      memoizedNodeText: nodeTexts
    };
  };

  // Builds a map of child leaf nodes by node id - memoized so that it only rebuilds the list if the data changes.
  const { memoizedLeavesById, memoizedAllLeaves, memoizedNodeText } = React.useMemo(() => memoizeNodes(data), [data]);
```
with `data` being passed as an argument to the memoization callback?

This is a genuine question as I'm not well versed with React memoization and I'm not sure how to really test this without a large dataset, I just saw that in the React docs they show usage as:

![image](https://user-images.githubusercontent.com/7115053/145070997-2d7e2850-3d74-4985-b94c-5b19b65a13eb.png)

Reply:
My understanding is that `React.useMemo` works like `React.useEffect` where it watches for changes in the dependencies listed in the dependency array at the end - and it only recalculates the value of the `memoizedValue` if those dependencies change.

Reply:
I misunderstood something from the docs, disregard :sweat_smile:

Reply:
@wise-king-sullyman I think both are correct. Defining a dedicated `memoizeNodes` function at render time just before passing useMemo an anonymous function that calls it behaves the same as defining that behavior directly in the anonymous function. The important part is the `[data]` dependency array which will prevent the whole anonymous function from being run again if the data didn't change. So it's a matter of preference - I think the example in the React docs was written that way for brevity.